flink/src/main/java/org/apache/iceberg/flink/source/RowDataRewriter.java

```diff
@@ -0,0 +1,193 @@
+    this.io = io;
+    this.encryptionManager = encryptionManager;
+
+    this.caseSensitive = caseSensitive;
+    this.nameMapping = table.properties().get(DEFAULT_NAME_MAPPING);
+
+    String formatString = table.properties().getOrDefault(
+        TableProperties.DEFAULT_FILE_FORMAT, TableProperties.DEFAULT_FILE_FORMAT_DEFAULT);
+    this.format = FileFormat.valueOf(formatString.toUpperCase(Locale.ENGLISH));
+
+    RowType flinkSchema = FlinkSchemaUtil.convert(table.schema());
+    taskWriterFactory = new RowDataTaskWriterFactory(table.schema(),
+        flinkSchema,
+        table.spec(),
+        table.locationProvider(),
+        table.io(),
+        table.encryption(),
+        Long.MAX_VALUE,
+        format,
+        table.properties());
+  }
+
+
+  public List<DataFile> rewriteDataForTasks(DataSet<CombinedScanTask> dataSet) throws Exception {
+
+    RewriteMap map = new RewriteMap(schema, nameMapping, io, caseSensitive,
+        encryptionManager,
+        taskWriterFactory);
+
+    DataSet<List<DataFile>> ds = dataSet.map(map);
+    return ds.collect().stream()
+        .flatMap(Collection::stream)
+        .collect(Collectors.toList());
+  }
+
+  public static class RewriteMap extends RichMapFunction<CombinedScanTask, List<DataFile>> {
+
+    private transient TaskWriter<RowData> writer;
+    private transient int subTaskId;
+    private transient int attemptId;
+
+    private final Schema schema;
+    private final String nameMapping;
+    private final FileIO io;
+    private final boolean caseSensitive;
+    private final EncryptionManager encryptionManager;
+    private final TaskWriterFactory<RowData> taskWriterFactory;
+
+    public RewriteMap(Schema schema, String nameMapping, FileIO io, boolean caseSensitive,
+                      EncryptionManager encryptionManager,
+                      TaskWriterFactory<RowData> taskWriterFactory) {
+      this.schema = schema;
+      this.nameMapping = nameMapping;
+      this.io = io;
+      this.caseSensitive = caseSensitive;
+      this.encryptionManager = encryptionManager;
+      this.taskWriterFactory = taskWriterFactory;
+    }
+
+    @Override
+    public void open(Configuration parameters) {
+      this.subTaskId = getRuntimeContext().getIndexOfThisSubtask();
+      this.attemptId = getRuntimeContext().getAttemptNumber();
+
+      // Initialize the task writer factory.
+      this.taskWriterFactory.initialize(subTaskId, attemptId);
+
+    }
+
+
+    @Override
+    public List<DataFile> map(CombinedScanTask task) throws Exception {
+      // Initialize the task writer.
+      this.writer = taskWriterFactory.create();
+      RowDataIterator iterator =
+          new RowDataIterator(task, io, encryptionManager, schema, schema,
+              nameMapping, caseSensitive);
+      try {
+        while (iterator.hasNext()) {
+          RowData rowData = iterator.next();
+          writer.write(rowData);
+        }
+        iterator.close();
+        writer.close();
```

Comment:
The following `writer.complete()` will close firstly and then return the completed data file list, pls see the javadoc in `TaskWriter`: 

```java
  /**
   * Close the writer and get the completed data files.
   *
   * @return the completed data files of this task writer.
   */
  DataFile[] complete() throws IOException;
```
So here we don't have to call `writer.close()` now.